File: src/main/scala/com/vc/hard/L871.scala
package com.vc.hard

object L871 {
  def minRefuelStops(target: Int, startFuel: Int, stations: Array[Array[Int]]): Int = {
    import scala.collection.mutable
    val pq = new mutable.PriorityQueue[Int]()
    var stop = startFuel
    val n = stations.length
    var i = 0
    var refuels = 0
    while(stop < target) {
      while(i < n && stop >= stations(i)(0)) {
        pq += stations(i)(1)
        i += 1
      }
      if(pq.isEmpty) return -1
      stop += pq.dequeue()
      refuels += 1
    }
    refuels
  }
}

File: src/main/java/com/vc/hard/L871.java
package com.vc.hard;

import java.util.Collections;
import java.util.PriorityQueue;

class L871 {
    public int minRefuelStops(int target, int startFuel, int[][] stations) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        int pos = startFuel;
        int index = 0;
        int refills = 0;
        while(pos < target) {
            while(index < stations.length && pos >= stations[index][0]) {
                pq.offer(stations[index][1]);
                index++;
            }
            if(pq.isEmpty()) return -1;
            pos += pq.poll();
            refills++;
        }
        return refills;
    }
}

